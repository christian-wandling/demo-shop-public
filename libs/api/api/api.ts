export * from './cart-items-api';
import { CartItemsApi } from './cart-items-api';
export * from './orders-api';
import { OrdersApi } from './orders-api';
export * from './products-api';
import { ProductsApi } from './products-api';
export * from './shopping-sessions-api';
import { ShoppingSessionsApi } from './shopping-sessions-api';
export * from './users-api';
import { UsersApi } from './users-api';
export const APIS = [CartItemsApi, OrdersApi, ProductsApi, ShoppingSessionsApi, UsersApi];
