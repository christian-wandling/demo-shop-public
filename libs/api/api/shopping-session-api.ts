/**
 * Demo Shop API
 *
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/* tslint:disable:no-unused-variable member-ordering */

import { Inject, Injectable, Optional } from '@angular/core';
import {
  HttpClient,
  HttpHeaders,
  HttpParams,
  HttpResponse,
  HttpEvent,
  HttpParameterCodec,
  HttpContext,
} from '@angular/common/http';
import { CustomHttpParameterCodec } from '../encoder';
import { Observable } from 'rxjs';

// @ts-ignore
import { AddCartItem401Response } from '../model/add-cart-item401-response';
// @ts-ignore
import { AddCartItemRequest } from '../model/add-cart-item-request';
// @ts-ignore
import { CartItemResponse } from '../model/cart-item-response';
// @ts-ignore
import { OrderResponse } from '../model/order-response';
// @ts-ignore
import { ShoppingSessionResponse } from '../model/shopping-session-response';
// @ts-ignore
import { UpdateCartItemQuantityRequest } from '../model/update-cart-item-quantity-request';
// @ts-ignore
import { UpdateCartItemQuantityResponse } from '../model/update-cart-item-quantity-response';
// @ts-ignore
import { ValidationProblemDetails } from '../model/validation-problem-details';

// @ts-ignore
import { BASE_PATH, COLLECTION_FORMATS } from '../variables';
import { Configuration } from '../configuration';

@Injectable({
  providedIn: 'root',
})
export class ShoppingSessionApi {
  protected basePath = 'http://localhost';
  public defaultHeaders = new HttpHeaders();
  public configuration = new Configuration();
  public encoder: HttpParameterCodec;

  constructor(
    protected httpClient: HttpClient,
    @Optional() @Inject(BASE_PATH) basePath: string | string[],
    @Optional() configuration: Configuration
  ) {
    if (configuration) {
      this.configuration = configuration;
    }
    if (typeof this.configuration.basePath !== 'string') {
      const firstBasePath = Array.isArray(basePath) ? basePath[0] : undefined;
      if (firstBasePath != undefined) {
        basePath = firstBasePath;
      }

      if (typeof basePath !== 'string') {
        basePath = this.basePath;
      }
      this.configuration.basePath = basePath;
    }
    this.encoder = this.configuration.encoder || new CustomHttpParameterCodec();
  }

  // @ts-ignore
  private addToHttpParams(httpParams: HttpParams, value: any, key?: string): HttpParams {
    if (typeof value === 'object' && value instanceof Date === false) {
      httpParams = this.addToHttpParamsRecursive(httpParams, value);
    } else {
      httpParams = this.addToHttpParamsRecursive(httpParams, value, key);
    }
    return httpParams;
  }

  private addToHttpParamsRecursive(httpParams: HttpParams, value?: any, key?: string): HttpParams {
    if (value == null) {
      return httpParams;
    }

    if (typeof value === 'object') {
      if (Array.isArray(value)) {
        (value as any[]).forEach(elem => (httpParams = this.addToHttpParamsRecursive(httpParams, elem, key)));
      } else if (value instanceof Date) {
        if (key != null) {
          httpParams = httpParams.append(key, (value as Date).toISOString().substring(0, 10));
        } else {
          throw Error('key may not be null if value is Date');
        }
      } else {
        Object.keys(value).forEach(
          k => (httpParams = this.addToHttpParamsRecursive(httpParams, value[k], key != null ? `${key}.${k}` : k))
        );
      }
    } else if (key != null) {
      httpParams = httpParams.append(key, value);
    } else {
      throw Error('key may not be null if value is not object or array');
    }
    return httpParams;
  }

  /**
   * Add cart item
   * Add a cart item to the shopping session of current user based on identity extracted from bearer token
   * @param addCartItemRequest
   * @param observe set whether or not to return the data Observable as the body, response or events. defaults to returning the body.
   * @param reportProgress flag to report request and response progress.
   */
  public addCartItem(
    addCartItemRequest?: AddCartItemRequest,
    observe?: 'body',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<CartItemResponse>;
  public addCartItem(
    addCartItemRequest?: AddCartItemRequest,
    observe?: 'response',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<HttpResponse<CartItemResponse>>;
  public addCartItem(
    addCartItemRequest?: AddCartItemRequest,
    observe?: 'events',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<HttpEvent<CartItemResponse>>;
  public addCartItem(
    addCartItemRequest?: AddCartItemRequest,
    observe: any = 'body',
    reportProgress: boolean = false,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<any> {
    let localVarHeaders = this.defaultHeaders;

    let localVarHttpHeaderAcceptSelected: string | undefined = options && options.httpHeaderAccept;
    if (localVarHttpHeaderAcceptSelected === undefined) {
      // to determine the Accept header
      const httpHeaderAccepts: string[] = ['text/plain', 'application/json', 'text/json'];
      localVarHttpHeaderAcceptSelected = this.configuration.selectHeaderAccept(httpHeaderAccepts);
    }
    if (localVarHttpHeaderAcceptSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Accept', localVarHttpHeaderAcceptSelected);
    }

    let localVarHttpContext: HttpContext | undefined = options && options.context;
    if (localVarHttpContext === undefined) {
      localVarHttpContext = new HttpContext();
    }

    let localVarTransferCache: boolean | undefined = options && options.transferCache;
    if (localVarTransferCache === undefined) {
      localVarTransferCache = true;
    }

    // to determine the Content-Type header
    const consumes: string[] = ['application/json', 'text/json', 'application/*+json'];
    const httpContentTypeSelected: string | undefined = this.configuration.selectHeaderContentType(consumes);
    if (httpContentTypeSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Content-Type', httpContentTypeSelected);
    }

    let responseType_: 'text' | 'json' | 'blob' = 'json';
    if (localVarHttpHeaderAcceptSelected) {
      if (localVarHttpHeaderAcceptSelected.startsWith('text')) {
        responseType_ = 'text';
      } else if (this.configuration.isJsonMime(localVarHttpHeaderAcceptSelected)) {
        responseType_ = 'json';
      } else {
        responseType_ = 'blob';
      }
    }

    let localVarPath = `/api/v1/shopping-sessions/current/cart-items`;
    return this.httpClient.request<CartItemResponse>('post', `${this.configuration.basePath}${localVarPath}`, {
      context: localVarHttpContext,
      body: addCartItemRequest,
      responseType: <any>responseType_,
      withCredentials: this.configuration.withCredentials,
      headers: localVarHeaders,
      observe: observe,
      transferCache: localVarTransferCache,
      reportProgress: reportProgress,
    });
  }

  /**
   * Checkout current shopping session
   * Check out by creating an order from the current shopping session and deleting the shopping session afterwardds
   * @param observe set whether or not to return the data Observable as the body, response or events. defaults to returning the body.
   * @param reportProgress flag to report request and response progress.
   */
  public checkout(
    observe?: 'body',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<OrderResponse>;
  public checkout(
    observe?: 'response',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<HttpResponse<OrderResponse>>;
  public checkout(
    observe?: 'events',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<HttpEvent<OrderResponse>>;
  public checkout(
    observe: any = 'body',
    reportProgress: boolean = false,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<any> {
    let localVarHeaders = this.defaultHeaders;

    let localVarHttpHeaderAcceptSelected: string | undefined = options && options.httpHeaderAccept;
    if (localVarHttpHeaderAcceptSelected === undefined) {
      // to determine the Accept header
      const httpHeaderAccepts: string[] = ['text/plain', 'application/json', 'text/json'];
      localVarHttpHeaderAcceptSelected = this.configuration.selectHeaderAccept(httpHeaderAccepts);
    }
    if (localVarHttpHeaderAcceptSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Accept', localVarHttpHeaderAcceptSelected);
    }

    let localVarHttpContext: HttpContext | undefined = options && options.context;
    if (localVarHttpContext === undefined) {
      localVarHttpContext = new HttpContext();
    }

    let localVarTransferCache: boolean | undefined = options && options.transferCache;
    if (localVarTransferCache === undefined) {
      localVarTransferCache = true;
    }

    let responseType_: 'text' | 'json' | 'blob' = 'json';
    if (localVarHttpHeaderAcceptSelected) {
      if (localVarHttpHeaderAcceptSelected.startsWith('text')) {
        responseType_ = 'text';
      } else if (this.configuration.isJsonMime(localVarHttpHeaderAcceptSelected)) {
        responseType_ = 'json';
      } else {
        responseType_ = 'blob';
      }
    }

    let localVarPath = `/api/v1/shopping-sessions/checkout`;
    return this.httpClient.request<OrderResponse>('post', `${this.configuration.basePath}${localVarPath}`, {
      context: localVarHttpContext,
      responseType: <any>responseType_,
      withCredentials: this.configuration.withCredentials,
      headers: localVarHeaders,
      observe: observe,
      transferCache: localVarTransferCache,
      reportProgress: reportProgress,
    });
  }

  /**
   * Remove cart item
   * Remove a cart item from the shopping session of current user based on identity extracted from bearer token
   * @param id
   * @param observe set whether or not to return the data Observable as the body, response or events. defaults to returning the body.
   * @param reportProgress flag to report request and response progress.
   */
  public removeCartItem(
    id: number,
    observe?: 'body',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<any>;
  public removeCartItem(
    id: number,
    observe?: 'response',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<HttpResponse<any>>;
  public removeCartItem(
    id: number,
    observe?: 'events',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<HttpEvent<any>>;
  public removeCartItem(
    id: number,
    observe: any = 'body',
    reportProgress: boolean = false,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<any> {
    if (id === null || id === undefined) {
      throw new Error('Required parameter id was null or undefined when calling removeCartItem.');
    }

    let localVarHeaders = this.defaultHeaders;

    let localVarHttpHeaderAcceptSelected: string | undefined = options && options.httpHeaderAccept;
    if (localVarHttpHeaderAcceptSelected === undefined) {
      // to determine the Accept header
      const httpHeaderAccepts: string[] = ['text/plain', 'application/json', 'text/json'];
      localVarHttpHeaderAcceptSelected = this.configuration.selectHeaderAccept(httpHeaderAccepts);
    }
    if (localVarHttpHeaderAcceptSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Accept', localVarHttpHeaderAcceptSelected);
    }

    let localVarHttpContext: HttpContext | undefined = options && options.context;
    if (localVarHttpContext === undefined) {
      localVarHttpContext = new HttpContext();
    }

    let localVarTransferCache: boolean | undefined = options && options.transferCache;
    if (localVarTransferCache === undefined) {
      localVarTransferCache = true;
    }

    let responseType_: 'text' | 'json' | 'blob' = 'json';
    if (localVarHttpHeaderAcceptSelected) {
      if (localVarHttpHeaderAcceptSelected.startsWith('text')) {
        responseType_ = 'text';
      } else if (this.configuration.isJsonMime(localVarHttpHeaderAcceptSelected)) {
        responseType_ = 'json';
      } else {
        responseType_ = 'blob';
      }
    }

    let localVarPath = `/api/v1/shopping-sessions/current/cart-items/${this.configuration.encodeParam({
      name: 'id',
      value: id,
      in: 'path',
      style: 'simple',
      explode: false,
      dataType: 'number',
      dataFormat: 'int32',
    })}`;
    return this.httpClient.request<any>('delete', `${this.configuration.basePath}${localVarPath}`, {
      context: localVarHttpContext,
      responseType: <any>responseType_,
      withCredentials: this.configuration.withCredentials,
      headers: localVarHeaders,
      observe: observe,
      transferCache: localVarTransferCache,
      reportProgress: reportProgress,
    });
  }

  /**
   * Resolve current shopping session
   * Resolve current shopping session based on identity extracted from bearer token
   * @param observe set whether or not to return the data Observable as the body, response or events. defaults to returning the body.
   * @param reportProgress flag to report request and response progress.
   */
  public resolveCurrentShoppingSession(
    observe?: 'body',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<ShoppingSessionResponse>;
  public resolveCurrentShoppingSession(
    observe?: 'response',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<HttpResponse<ShoppingSessionResponse>>;
  public resolveCurrentShoppingSession(
    observe?: 'events',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<HttpEvent<ShoppingSessionResponse>>;
  public resolveCurrentShoppingSession(
    observe: any = 'body',
    reportProgress: boolean = false,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<any> {
    let localVarHeaders = this.defaultHeaders;

    let localVarHttpHeaderAcceptSelected: string | undefined = options && options.httpHeaderAccept;
    if (localVarHttpHeaderAcceptSelected === undefined) {
      // to determine the Accept header
      const httpHeaderAccepts: string[] = ['text/plain', 'application/json', 'text/json'];
      localVarHttpHeaderAcceptSelected = this.configuration.selectHeaderAccept(httpHeaderAccepts);
    }
    if (localVarHttpHeaderAcceptSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Accept', localVarHttpHeaderAcceptSelected);
    }

    let localVarHttpContext: HttpContext | undefined = options && options.context;
    if (localVarHttpContext === undefined) {
      localVarHttpContext = new HttpContext();
    }

    let localVarTransferCache: boolean | undefined = options && options.transferCache;
    if (localVarTransferCache === undefined) {
      localVarTransferCache = true;
    }

    let responseType_: 'text' | 'json' | 'blob' = 'json';
    if (localVarHttpHeaderAcceptSelected) {
      if (localVarHttpHeaderAcceptSelected.startsWith('text')) {
        responseType_ = 'text';
      } else if (this.configuration.isJsonMime(localVarHttpHeaderAcceptSelected)) {
        responseType_ = 'json';
      } else {
        responseType_ = 'blob';
      }
    }

    let localVarPath = `/api/v1/shopping-sessions/current`;
    return this.httpClient.request<ShoppingSessionResponse>('post', `${this.configuration.basePath}${localVarPath}`, {
      context: localVarHttpContext,
      responseType: <any>responseType_,
      withCredentials: this.configuration.withCredentials,
      headers: localVarHeaders,
      observe: observe,
      transferCache: localVarTransferCache,
      reportProgress: reportProgress,
    });
  }

  /**
   * Update cart item quantity
   * Update quantity of a cart item in shopping session of current user based on identity extracted from bearer token
   * @param id
   * @param updateCartItemQuantityRequest
   * @param observe set whether or not to return the data Observable as the body, response or events. defaults to returning the body.
   * @param reportProgress flag to report request and response progress.
   */
  public updateCartItemQuantity(
    id: number,
    updateCartItemQuantityRequest: UpdateCartItemQuantityRequest,
    observe?: 'body',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<UpdateCartItemQuantityResponse>;
  public updateCartItemQuantity(
    id: number,
    updateCartItemQuantityRequest: UpdateCartItemQuantityRequest,
    observe?: 'response',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<HttpResponse<UpdateCartItemQuantityResponse>>;
  public updateCartItemQuantity(
    id: number,
    updateCartItemQuantityRequest: UpdateCartItemQuantityRequest,
    observe?: 'events',
    reportProgress?: boolean,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<HttpEvent<UpdateCartItemQuantityResponse>>;
  public updateCartItemQuantity(
    id: number,
    updateCartItemQuantityRequest: UpdateCartItemQuantityRequest,
    observe: any = 'body',
    reportProgress: boolean = false,
    options?: {
      httpHeaderAccept?: 'text/plain' | 'application/json' | 'text/json';
      context?: HttpContext;
      transferCache?: boolean;
    }
  ): Observable<any> {
    if (id === null || id === undefined) {
      throw new Error('Required parameter id was null or undefined when calling updateCartItemQuantity.');
    }
    if (updateCartItemQuantityRequest === null || updateCartItemQuantityRequest === undefined) {
      throw new Error(
        'Required parameter updateCartItemQuantityRequest was null or undefined when calling updateCartItemQuantity.'
      );
    }

    let localVarHeaders = this.defaultHeaders;

    let localVarHttpHeaderAcceptSelected: string | undefined = options && options.httpHeaderAccept;
    if (localVarHttpHeaderAcceptSelected === undefined) {
      // to determine the Accept header
      const httpHeaderAccepts: string[] = ['text/plain', 'application/json', 'text/json'];
      localVarHttpHeaderAcceptSelected = this.configuration.selectHeaderAccept(httpHeaderAccepts);
    }
    if (localVarHttpHeaderAcceptSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Accept', localVarHttpHeaderAcceptSelected);
    }

    let localVarHttpContext: HttpContext | undefined = options && options.context;
    if (localVarHttpContext === undefined) {
      localVarHttpContext = new HttpContext();
    }

    let localVarTransferCache: boolean | undefined = options && options.transferCache;
    if (localVarTransferCache === undefined) {
      localVarTransferCache = true;
    }

    // to determine the Content-Type header
    const consumes: string[] = ['application/json', 'text/json', 'application/*+json'];
    const httpContentTypeSelected: string | undefined = this.configuration.selectHeaderContentType(consumes);
    if (httpContentTypeSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Content-Type', httpContentTypeSelected);
    }

    let responseType_: 'text' | 'json' | 'blob' = 'json';
    if (localVarHttpHeaderAcceptSelected) {
      if (localVarHttpHeaderAcceptSelected.startsWith('text')) {
        responseType_ = 'text';
      } else if (this.configuration.isJsonMime(localVarHttpHeaderAcceptSelected)) {
        responseType_ = 'json';
      } else {
        responseType_ = 'blob';
      }
    }

    let localVarPath = `/api/v1/shopping-sessions/current/cart-items/${this.configuration.encodeParam({
      name: 'id',
      value: id,
      in: 'path',
      style: 'simple',
      explode: false,
      dataType: 'number',
      dataFormat: 'int32',
    })}`;
    return this.httpClient.request<UpdateCartItemQuantityResponse>(
      'patch',
      `${this.configuration.basePath}${localVarPath}`,
      {
        context: localVarHttpContext,
        body: updateCartItemQuantityRequest,
        responseType: <any>responseType_,
        withCredentials: this.configuration.withCredentials,
        headers: localVarHeaders,
        observe: observe,
        transferCache: localVarTransferCache,
        reportProgress: reportProgress,
      }
    );
  }
}
