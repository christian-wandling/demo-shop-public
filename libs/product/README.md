# `@demo-shop/product`

A library for product management. It enables displaying and filtering the product catalog and add them to the shopping cart.

## Features

- **Product Catalog**

  - Response grid view of product items

- **Product Details**

  - Detailed product information

- **Shopping Integration**

  - One-click add to cart functionality

- **Product Filtering**

  - Realtime full text search filter
  - Supports text and category filtering

- **State Management**
  - Signal-based reactive store
  - Synchronization with backend

## API Reference

For a complete API reference, see the [Compodoc documentation](https://christian-wandling.github.io/demo-shop-public/compodoc/frontend).

## Running unit tests

Run `nx test product` to execute the unit tests.
