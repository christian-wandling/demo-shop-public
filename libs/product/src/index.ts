export * from './lib/product.routes';
export * from './lib/models/product-filter';
export * from './lib/product.facade';
export * from './lib/components/product-search/product-search.component';
