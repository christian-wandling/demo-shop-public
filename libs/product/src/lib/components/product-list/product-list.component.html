<div class="mx-auto max-w-2xl px-4 py-16 sm:px-6 sm:py-24 lg:max-w-7xl lg:px-8">
  <h2 class="sr-only">Products</h2>

  <div class="grid grid-cols-1 gap-x-6 gap-y-10 sm:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4 xl:gap-x-8">
    @for (product of products(); track product.id) {
    <a
      routerLink="/products/{{ product.id }}"
      class="product-item group"
      [attr.data-testid]="'product-list-item-link' + product.id">
      <div class="aspect-h-1 aspect-w-1 w-full overflow-hidden rounded-lg bg-gray-200 xl:aspect-h-8 xl:aspect-w-7">
        <img
          [ngSrc]="product.thumbnail.uri"
          [alt]="product.name"
          placeholder
          class="h-full w-full object-cover object-center group-hover:opacity-75"
          [attr.data-testid]="'product-list-item-image' + product.id"
          fill="" />
      </div>
      <h3 class="mt-4 text-sm text-gray-700" [attr.data-testid]="'product-list-item-name' + product.id">
        {{ product.name }}
      </h3>
      <p class="mt-1 text-lg font-medium text-gray-900" [attr.data-testid]="'product-list-item-price' + product.id">
        {{ product.price | currency }}
      </p>
    </a>
    }
  </div>
</div>
