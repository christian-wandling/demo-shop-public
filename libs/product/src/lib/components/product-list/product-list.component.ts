import { ChangeDetectionStrategy, Component, inject, OnInit } from '@angular/core';
import { CommonModule, NgOptimizedImage } from '@angular/common';
import { ProductFacade } from '../../product.facade';
import { RouterLink } from '@angular/router';

/**
 * Component for displaying a list of products.
 *
 * This standalone component fetches and displays a list of products,
 * using the ProductFacade for data access and OnPush change detection for performance.
 *
 * @example
 * <lib-product-list/>
 */
@Component({
  selector: 'lib-product-list',
  standalone: true,
  imports: [CommonModule, NgOptimizedImage, RouterLink],
  templateUrl: './product-list.component.html',
  styleUrl: './product-list.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush,
  host: {
    'data-testid': 'product-list',
    class: 'bg-white',
  },
})
export class ProductListComponent implements OnInit {
  readonly #productFacade = inject(ProductFacade);

  readonly products = this.#productFacade.getFiltered();

  /**
   * Fetches all products from the backend.
   */
  ngOnInit() {
    this.#productFacade.fetchAll();
  }
}
