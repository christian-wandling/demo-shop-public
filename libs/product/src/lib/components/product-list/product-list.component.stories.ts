import { applicationConfig, Meta, StoryObj } from '@storybook/angular';
import { CommonModule, NgOptimizedImage } from '@angular/common';
import { provideRouter } from '@angular/router';
import { importProvidersFrom, signal } from '@angular/core';
import { ProductFacade } from '../../product.facade';
import { within } from '@storybook/testing-library';
import { expect } from '@storybook/jest';
import { ProductListComponent } from './product-list.component';
import { provideImageLoader } from '@demo-shop/shared';
import { mockProducts } from '../../+mock/mock-products';

const mockProductFacade = {
  getFiltered: () => signal(mockProducts),
  fetchAll: () => {
    return;
  },
};

const meta: Meta<ProductListComponent> = {
  component: ProductListComponent,
  title: 'Product/ProductListComponent',
  decorators: [
    applicationConfig({
      providers: [
        importProvidersFrom(CommonModule, NgOptimizedImage),
        provideImageLoader(),
        provideRouter([{ path: '**', redirectTo: '' }]),
        { provide: ProductFacade, useValue: mockProductFacade },
      ],
    }),
  ],
  parameters: {
    controls: {
      exclude: ['#productFacade', 'products', 'ngOnInit'],
    },
  },
};
export default meta;
type Story = StoryObj<ProductListComponent>;

export const Default: Story = {
  play: async ({ canvasElement, args }) => {
    const firstImage: HTMLImageElement = within(canvasElement).getByAltText(mockProducts[0].name);
    const images: HTMLImageElement[] = within(canvasElement).getAllByRole('img');
    expect(firstImage).toBeTruthy();
    expect(images).toHaveLength(mockProducts.length);
  },
};
