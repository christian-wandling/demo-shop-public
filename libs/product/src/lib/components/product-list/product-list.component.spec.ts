import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ProductListComponent } from './product-list.component';
import { ProductFacade } from '../../product.facade';
import { ProductResponse } from '@demo-shop/api';
import { signal } from '@angular/core';
import { provideRouter } from '@angular/router';
import { By } from '@angular/platform-browser';

describe('ProductListComponent', () => {
  let component: ProductListComponent;
  let fixture: ComponentFixture<ProductListComponent>;

  const mockProducts: ProductResponse[] = [
    {
      id: 1,
      name: 'Product 1',
      price: 10,
      description: '',
      categories: [],
      images: [
        {
          name: 'name',
          uri: 'uri',
        },
      ],
      thumbnail: {
        name: 'name',
        uri: 'uri',
      },
    },
    {
      id: 2,
      name: 'Product 2',
      price: 20,
      description: '',
      categories: [],
      images: [
        {
          name: 'name',
          uri: 'uri',
        },
      ],
      thumbnail: {
        name: 'name',
        uri: 'uri',
      },
    },
  ];

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [ProductListComponent],
      providers: [
        {
          provide: ProductFacade,
          useValue: {
            getFiltered: jest.fn().mockReturnValue(signal(mockProducts)),
            fetchAll: jest.fn(),
          },
        },
        provideRouter([]),
      ],
    }).compileComponents();

    fixture = TestBed.createComponent(ProductListComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should render the products', () => {
    const productItems = fixture.debugElement.queryAll(By.css('.product-item'));

    expect(productItems.length).toBe(mockProducts.length);
  });

  it('should match the snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
