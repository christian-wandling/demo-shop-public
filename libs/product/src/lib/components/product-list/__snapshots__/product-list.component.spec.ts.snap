// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductListComponent should match the snapshot 1`] = `
<lib-product-list
  products={[Function Function]}
>
  <div
    class="mx-auto max-w-2xl px-4 py-16 sm:px-6 sm:py-24 lg:max-w-7xl lg:px-8"
  >
    <h2
      class="sr-only"
    >
      Products
    </h2>
    <div
      class="grid grid-cols-1 gap-x-6 gap-y-10 sm:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4 xl:gap-x-8"
    >
      <a
        class="product-item group"
        href="/products/1"
      >
        <div
          class="aspect-h-1 aspect-w-1 w-full overflow-hidden rounded-lg bg-gray-200 xl:aspect-h-8 xl:aspect-w-7"
        >
          <img
            alt="Product 1"
            class="h-full w-full object-cover object-center group-hover:opacity-75"
            fetchpriority="auto"
            fill=""
            loading="lazy"
            ng-img="true"
            onerror="this.src='images/placeholder-image.jpg'"
            sizes="100vw"
            src="uri"
            style="position: absolute; width: 100%; height: 100%; inset: 0;"
          />
        </div>
        <h3
          class="mt-4 text-sm text-gray-700"
        >
          Product 1
        </h3>
        <p
          class="mt-1 text-lg font-medium text-gray-900"
        >
          $10.00
        </p>
      </a>
      <a
        class="product-item group"
        href="/products/2"
      >
        <div
          class="aspect-h-1 aspect-w-1 w-full overflow-hidden rounded-lg bg-gray-200 xl:aspect-h-8 xl:aspect-w-7"
        >
          <img
            alt="Product 2"
            class="h-full w-full object-cover object-center group-hover:opacity-75"
            fetchpriority="auto"
            fill=""
            loading="lazy"
            ng-img="true"
            onerror="this.src='images/placeholder-image.jpg'"
            sizes="100vw"
            src="uri"
            style="position: absolute; width: 100%; height: 100%; inset: 0;"
          />
        </div>
        <h3
          class="mt-4 text-sm text-gray-700"
        >
          Product 2
        </h3>
        <p
          class="mt-1 text-lg font-medium text-gray-900"
        >
          $20.00
        </p>
      </a>
    </div>
  </div>
</lib-product-list>
`;
